File: shared/image-paths.js
// shared/image-paths.js — master id → icons/ 相对路径
//
// 资源位置: crawl/icons/ (从 D:\bxb 拷过来、见 scripts/master_to_business/copy_images.py)
// .gitignore 排除、~150MB 不入 git
//
// 命名规则:
// - chara : icons/chara/{weapons.id 6位}.png
// - masou : icons/masou/{weapon_costumes.id 7位}.png
// - crystal: icons/crystal/{materials.id}_{1或2}.png
// - bg    : icons/bg/{pictures.id}.png
// - soul  : icons/soul/{jobs.texture_id}.png  ← 用 texture_id 不是 jobs.id
//
// browser 端不能 fs check、用 <img onerror="this.style.display='none'"> fallback text-only

const BASE = '../icons';

export const charaIcon = (variantId) => `${BASE}/chara/${variantId}.png`;
export const masouIcon = (costumeId) => `${BASE}/masou/${costumeId}.png`;
export const bgIcon = (pictureId) => `${BASE}/bg/${pictureId}.png`;
export const soulIcon = (textureId) => `${BASE}/soul/${textureId}.png`;

// 注: 曾有 crystalIcon (multi-suffix `{id}_{N}.png`、onerror fallback _2) — 已删。
// D:/bxb 重解后无 _N 源、cascade 不再需要;crystal 图标路径现在由
// shared/crystal-spec.js 直接拼 `../icons/crystal/{id}.png`。

// chara icon 叠层资源 (尺寸都跟 chara icon 100×100 配合好):
// - marriage 框 100×100 (=base)、level 1/2/3 红→粉→金渐变
// - weapon_type_42 42×42 (左上角)
// - element_list 33×33 (右上角)
const marriageOverlay = (level = 2) => `${BASE}/_misc/marriage_${level}.png`;
const weaponTypeIcon = (weaponTypeId) => `${BASE}/_app_icons/icon_weapon_type_42_${weaponTypeId}.png`;
const elementIcon = (elementId) => `${BASE}/_app_icons/icon_element_list_${elementId}.png`;

// chara icon 叠层 HTML 生成器 — 4 层 z-index: base / marriage / type / element
// 用法:
//   container.innerHTML = charaIconStack({
//     variantId: c.id, name: c.name,
//     elementId: c.element, weaponTypeId: c.weapon,
//     marriageLevel: 2,  // 0 = 不叠 marriage、1/2/3 = 红/粉/金
//   });
// lazy 模式:
//   'native' (默认): `loading="lazy"`、native HTML5 lazy、适用 document-scroll 场景 (slot card 等)
//   'io': src→data-src、caller 需调 setupLazyImg(scrollRoot)、适用自定义 scroll 容器 (modal list 等)
export const charaIconStack = ({
  variantId, name = '', elementId, weaponTypeId, marriageLevel = 0, className = 'chara-thumb',
  lazy = 'native',
}) => {
  const safe = String(name).replace(/"/g, '&quot;');
  const onerr = `onerror="this.style.visibility='hidden'"`;
  const srcAttr = (s) => lazy === 'io' ? `data-src="${s}"` : `src="${s}" loading="lazy"`;
  const layers = [
    `<img class="ct-base" ${srcAttr(charaIcon(variantId))} alt="${safe}" ${onerr}>`,
  ];
  if (marriageLevel) {
    layers.push(`<img class="ct-marriage" ${srcAttr(marriageOverlay(marriageLevel))} alt="" ${onerr}>`);
  }
  if (weaponTypeId) {
    layers.push(`<img class="ct-type" ${srcAttr(weaponTypeIcon(weaponTypeId))} alt="" ${onerr}>`);
  }
  if (elementId) {
    layers.push(`<img class="ct-element" ${srcAttr(elementIcon(elementId))} alt="" ${onerr}>`);
  }
  return `<div class="${className}">${layers.join('')}</div>`;
};

// 注: 曾有 imgWithFallback (通用 <img> 生成器 + onerror 隐藏) — 0 callsite,已删。
// 各 viewer 自己拼 <img>(通常还要带 data-src 走 shared/lazy-img.js 的容器内懒加载,
// 这个助手写死 loading="lazy" 反而不适用)。

File: scripts/master_to_business/image_paths.py
"""image_paths.py — master id → 本地 icon path 解析。

资源位置: crawl/icons/ (从 D:\bxb 拷过来、见 copy_images.py)。
crawl/icons/ 在 .gitignore 排除、~150MB 不入 git。

命名规则:
- chara : icons/chara/{weapons.id 6位}.png           99.8%
- masou : icons/masou/{weapon_costumes.id 7位}.png   99%
- crystal: icons/crystal/{materials.id}_{1或2}.png   99.5% (multi-suffix 时优先 _1)
- bg    : icons/bg/{pictures.id}.png                 98.6%
- soul  : icons/soul/{jobs.texture_id}.png           99% (用 texture_id、不是 jobs.id)

fallback: file 不存在 → 返回 None、前端 viewer text-only 渲染。

用法:
    from image_paths import chara_icon_path
    p = chara_icon_path(100101)
    if p: <img src=p>           # p 是相对项目 root 的路径如 'icons/chara/100101.png'
    else: <span>{name}</span>
"""
from pathlib import Path

PROJECT_ROOT = Path(__file__).resolve().parents[2]
ICONS_DIR = PROJECT_ROOT / "icons"

CHARA_DIR = ICONS_DIR / "chara"
MASOU_DIR = ICONS_DIR / "masou"
CRYSTAL_DIR = ICONS_DIR / "crystal"
BG_DIR = ICONS_DIR / "bg"
SOUL_DIR = ICONS_DIR / "soul"


def _rel_or_none(p: Path):
    """返回相对 PROJECT_ROOT 的路径 string (含 / 分隔、前端友好)、不存在返回 None"""
    if not p.is_file():
        return None
    return p.relative_to(PROJECT_ROOT).as_posix()


def chara_icon_path(variant_id: int):
    """chara variant (weapons.id 6 位) → icons/chara/{variant_id}.png"""
    return _rel_or_none(CHARA_DIR / f"{variant_id}.png")


def masou_icon_path(costume_id: int):
    """masou (weapon_costumes.id 7 位) → icons/masou/{costume_id}.png"""
    return _rel_or_none(MASOU_DIR / f"{costume_id}.png")


def crystal_icon_path(material_id: int):
    """crystal (materials.id) → icons/crystal/{id}_{1或2}.png
    多 suffix 时优先 _1、不存在再 _2"""
    for sfx in (1, 2, 3, 4):
        p = CRYSTAL_DIR / f"{material_id}_{sfx}.png"
        if p.is_file():
            return p.relative_to(PROJECT_ROOT).as_posix()
    return None


def bg_icon_path(picture_id: int):
    """bladegraph (pictures.id) → icons/bg/{picture_id}.png"""
    return _rel_or_none(BG_DIR / f"{picture_id}.png")


def soul_icon_path(texture_id: int):
    """soul → icons/soul/{jobs.texture_id}.png
    注: 用 jobs.texture_id (不是 jobs.id)"""
    return _rel_or_none(SOUL_DIR / f"{texture_id}.png")


# 自检 + 统计 cover 率
if __name__ == "__main__":
    import json
    import sys
    sys.path.insert(0, str(Path(__file__).resolve().parent))
    from paths import MASTER_DIR

    weapons = json.loads((MASTER_DIR / "weapons.json").read_text(encoding="utf-8"))
    materials = json.loads((MASTER_DIR / "materials.json").read_text(encoding="utf-8"))
    pictures = json.loads((MASTER_DIR / "pictures.json").read_text(encoding="utf-8"))
    jobs = json.loads((MASTER_DIR / "jobs.json").read_text(encoding="utf-8"))
    wcs = json.loads((MASTER_DIR / "weapon_costumes.json").read_text(encoding="utf-8"))

    print(f"=== image_paths cover 统计 ===")
    print(f"icons/ root: {ICONS_DIR.is_dir()}")

    n = sum(1 for w in weapons if chara_icon_path(w["id"]))
    print(f"chara : {n} / {len(weapons)} ({100*n/len(weapons):.1f}%)")

    n = sum(1 for w in wcs if masou_icon_path(w["id"]))
    print(f"masou : {n} / {len(wcs)} ({100*n/len(wcs):.1f}%)")

    n = sum(1 for m in materials if crystal_icon_path(m["id"]))
    print(f"crystal: {n} / {len(materials)} ({100*n/len(materials):.1f}%)")

    n = sum(1 for p in pictures if bg_icon_path(p["id"]))
    print(f"bg    : {n} / {len(pictures)} ({100*n/len(pictures):.1f}%)")

    n = sum(1 for j in jobs if j.get("texture_id") and soul_icon_path(j["texture_id"]))
    print(f"soul  : {n} / {len(jobs)} ({100*n/len(jobs):.1f}%)")
